python/tvm/tir/transform/inject_rolling_buffer.py

```diff
@@ -0,0 +1,214 @@
+# under the License.
+"""Inject rolling buffers through a TIR transformation."""
+# pylint: disable=invalid-name,unused-argument,inconsistent-return-statements
+from collections import defaultdict
+
+import tvm
+from tvm import arith
+
+
+def InjectRollingBuffer():
+    """Inject rolling buffer statements.
+
+    Returns
+    -------
+    fpass : tvm.transform.Pass
+        The pass
+    """
+    buffer_to_attrs = defaultdict(list)
+    rolling_buffers = set()
+    rolling_buffer_to_info = dict()
+    iter_vars = list()
+    hoist_buffer_to_for = defaultdict(list)
+
+    class RollingBufferInfo:
+        def __init__(self, rolling_axis, rolling_extent, axis_overlaps, axis_iter_vars):
+            self.rolling_axis = rolling_axis
+            self.rolling_extent = rolling_extent
+            self.axis_overlaps = axis_overlaps
+            self.axis_iter_vars = axis_iter_vars
+
+    def _pre_visit(stmt):
+        if isinstance(stmt, tvm.tir.For):
+            # Manage the stack of iter_vars
+            iter_vars.append(stmt)
+
+        elif isinstance(stmt, tvm.tir.AttrStmt):
+            if isinstance(stmt.node, tvm.tir.Buffer):
+                if stmt.attr_key == "rolling_buffer" and stmt.value.value:
+                    # If the attribute is indicating that a buffer should be a rolling
+                    # buffer, then update the rolling_buffers set to include the bufffer
+                    rolling_buffers.add(stmt.node)
+                # Keep a dictionary associating attribute statements with the buffers
+                # they reference. We'll need this if the buffer gets hoisted and we
+                # need to hoist all of its attributes at the same time.
+                buffer_to_attrs[stmt.node].append(stmt)
+
+        elif isinstance(stmt, tvm.tir.BufferRealize):
+            if stmt.buffer in rolling_buffers:
+                # If a BufferRealize has been identified as needing to be made into
+                # a rolling buffer, begin the analysis...
+                bound_iter_vars = []
+                bound_strides = []
+                bound_overlaps = []
+                # We use the bound information of the BufferRealize to calculate
+                # how we can legally roll
+                for bound in stmt.bounds:
+                    # If the bound is an int, we can't roll over it
+                    if isinstance(bound.min, tvm.tir.IntImm):
+                        iter_var = None
+                        stride = 0
+                    # If the bound is just a Var, that implies the stride is 1
+                    elif isinstance(bound.min, tvm.tir.Var):
+                        iter_var = bound.min
+                        stride = 1
+                    # Otherwise, it's the iter var multiplied by the stride
+                    # If not we're in unknown behaviour, so assert
+                    else:
+                        assert isinstance(bound.min, tvm.tir.Mul)
```

Comment:
Should we simply not unrolling instead of asserting?

Reply:
I see how it would be hard to stop the pass in the middle 

Depending on the confidence, this should be considered to be an Exception.
To be safe, we can raise an exception ?

Given the immutability of TVM objects, I think we might need to "try" running these sort of passes -- but that part might be out of scope for this PR.

Reply:
I couldn't think of an especially clean way to handle this. My primary reference for this is `inject_double_buffer.cc` and there the behaviour is to CHECK assert when the pass fails. I'm not sure exceptions are necessarily a better way to go here because that suggests to a degree that they can be handled. For now I've compromised by keeping the asserts but adding an assert message to at least explain what went wrong. Let me know what you think.

Reply:
Well; the reason I suggested exceptions were to transfer the responsibility to the caller (i.e., pass pipeline) to see if the pass is successful. Therefore the "handling" is not running a failing pass by the caller. However, given the status of catching TVM exceptions (last time I tried this, could not catch exception by type), I think assert might also work.